File: android/src/main/java/com/lightningkite/khrysalis/net/HttpClient.websocket.actual.kt
package com.lightningkite.khrysalis.net

import com.lightningkite.khrysalis.bytes.Data
import com.lightningkite.khrysalis.net.HttpClient.threadCorrectly
import io.reactivex.Observable
import io.reactivex.Single
import io.reactivex.disposables.Disposable
import io.reactivex.subjects.PublishSubject
import okhttp3.*
import okio.ByteString
import java.util.*


fun HttpClient.webSocket(
    url: String,
    headers: Map<String, String> = mapOf()
): Single<ConnectedWebSocket> = Single.create<ConnectedWebSocket> { emitter ->
    val out = PublishSubject.create<WebSocketFrame>()
    var external: ConnectedWebSocketImpl? = null
    val listener = object : WebSocketListener() {
        override fun onOpen(webSocket: WebSocket, response: Response) {
            println("Socket to $url opened successfully.")
            if (external == null) {
                external = ConnectedWebSocketImpl(webSocket, out.threadCorrectly())
                emitter.onSuccess(external!!)
            }
        }

        override fun onFailure(webSocket: WebSocket, t: Throwable, response: Response?) {
            println("Socket to $url failed with $t.")
            val external = external
            if (external == null) {
                emitter.onError(t)
            } else {
                out.onError(t)
                external.dispose()
            }
        }

        override fun onClosing(webSocket: WebSocket, code: Int, reason: String) {
            println("Socket to $url closing.")
            external?.let {
                out.onComplete()
                it.dispose()
            }
        }

        override fun onMessage(webSocket: WebSocket, text: String) {
            println("Socket to $url got message '$text'.")
            out.onNext(WebSocketFrame(text = text))
        }

        override fun onMessage(webSocket: WebSocket, bytes: ByteString) {
            println("Socket to $url got binary message of length ${bytes.size()}.")
            out.onNext(WebSocketFrame(binary = bytes.toByteArray()))
        }

        override fun onClosed(webSocket: WebSocket, code: Int, reason: String) {
            if (external == null) {
                emitter.onError(Exception("Socket closed before release."))
            }
            println("Socket to $url closed.")
        }
    }
    client.newWebSocket(
        Request.Builder()
            .url(url.replace("http", "ws"))
            .headers(Headers.of(headers))
            .addHeader("Accept-Language", Locale.getDefault().language)
            .build(),
        listener
    )
}.threadCorrectly()

private class ConnectedWebSocketImpl(val socket: WebSocket, override val read: Observable<WebSocketFrame>) :
    ConnectedWebSocket {
    override fun writeBinary(binary: Data) {
        println("Socket sent binary data of length ${binary.size}")
        socket.send(ByteString.of(binary, 0, binary.size))
    }

    override fun writeText(text: String) {
        println("Socket sent text '$text'.")
        socket.send(text)
    }

    var closed: Boolean = false
    override fun isDisposed(): Boolean {
        return closed
    }

    override fun dispose() {
        println("Socket requested disconnect locally.")
        socket.close(1000, null)
        closed = true
    }

}

class WebSocketFrame(val binary: Data? = null, val text: String? = null) {
    override fun toString(): String {
        return text ?: binary?.toString(Charsets.UTF_8) ?: "<Empty Frame>"
    }
}

File: ios/Khrysalis/com/lightningkite/khrysalis/net/HttpClient.websocket.actual.swift
import RxSwift
import RxRelay
import Starscream

//--- HttpClient.webSocket(String, Map<String,String>)
public extension HttpClient {
    static func webSocket(_ url: String, _ headers: Dictionary<String, String> = [:]) -> Observable<ConnectedWebSocket> {
        return Observable.using({ () -> ConnectedWebSocket in
            var out = ConnectedWebSocket(url: url)
            var request = URLRequest(url: URL(string: url)!)
            for (key, value) in headers {
                request.setValue(value, forHTTPHeaderField: key)
            }
            let socket = WebSocket(request: request)
            socket.delegate = out
            out.underlyingSocket = socket
            socket.connect()
            return out
        }, observableFactory: { $0.ownConnection })
    }
    static func webSocket(url: String, headers: Dictionary<String, String> = [:]) -> Observable<ConnectedWebSocket> {
        return webSocket(url, headers)
    }
}

//--- ConnectedWebSocket.{
public final class ConnectedWebSocket: WebSocketDelegate, Disposable {
    
    
    //--- ConnectedWebSocket.ownConnection
    public var ownConnection = PublishSubject<ConnectedWebSocket>.create()
    var underlyingSocket: WebSocket?
    var url: String
    //--- ConnectedWebSocket.read
    private let _read: PublishSubject<WebSocketFrame> = PublishSubject.create()
    public var read: Observable<WebSocketFrame> { return _read }
    
    //--- ConnectedWebSocket.Primary Constructor
    init(url: String) {
        self.url = url
    }
    
    
    //--- ConnectedWebSocket.onComplete()
    public func onComplete() -> Void {
        underlyingSocket?.disconnect(closeCode: 1000)
    }
    
    //--- ConnectedWebSocket.onNext(WebSocketFrame)
    public func onNext(_ t: WebSocketFrame) -> Void {
        if let text = t.text {
            underlyingSocket?.write(string: text, completion: nil)
        }
        if let binary = t.binary {
            underlyingSocket?.write(data: binary, completion: nil)
        }
    }
    public func onNext(t: WebSocketFrame) -> Void {
        return onNext(t)
    }
    
    //--- ConnectedWebSocket.onError(Throwable)
    public func onError(_ e: Throwable) -> Void {
        underlyingSocket?.disconnect(closeCode: 1011)
    }
    public func onError(e: Throwable) -> Void {
        return onError(e)
    }
    
    //--- Receive
    public func didReceive(event: WebSocketEvent, client: WebSocket) {
        switch event {
        case .binary(let data):
            print("Socket to \(url) got binary message of length '\(data.count)'")
            _read.onNext(WebSocketFrame(binary: data))
            break
        case .text(let string):
            print("Socket to \(url) got message '\(string)'")
            _read.onNext(WebSocketFrame(text: string))
            break
        case .connected(let headers):
            print("Socket to \(url) opened successfully.")
            ownConnection.onNext(self)
            break
        case .disconnected(let reason, let code):
            print("Socket to \(url) disconnecting with code \(code). Reason: \(reason)")
            ownConnection.onCompleted()
            _read.onComplete()
            break
        case .error(let error):
            print("Socket to \(url) failed with error \(error)")
            ownConnection.onError(error ?? Exception())
            read.onError(error ?? Exception())
            break
        case .cancelled:
            print("Socket to \(url) cancelled")
            ownConnection.onError(Exception("Socket connection cancelled."))
            _read.onComplete()
            break
        default:
            break
        }
    }
    
    public func dispose() {
        print("Socket to \(url) was disposed, closing with OK code.")
        underlyingSocket?.disconnect(closeCode: 1000)
        ownConnection.onCompleted()
        _read.onComplete()
    }
    
    //--- ConnectedWebSocket.}
}

//--- WebSocketFrame.{
public struct WebSocketFrame: CustomStringConvertible {
    
    public let binary: Data?
    public let text: String?
    
    //--- WebSocketFrame.Primary Constructor
    public init(binary: Data? = nil, text: String? = nil) {
        self.binary = binary
        self.text = text
    }
    
    //--- WebSocketFrame.toString()
    public var description: String {
        return text ?? "<binary>"
    }
    
    //--- WebSocketFrame.}
}
